src/Std/Data/HashSet/Lemmas.lean

```diff
@@ -0,0 +1,274 @@
+  HashMap.contains_insert
+
+@[simp]
+theorem mem_insert [EquivBEq α] [LawfulHashable α] {k a : α} : a ∈ m.insert k ↔ a == k ∨ a ∈ m :=
+  HashMap.mem_insert
+
+theorem contains_of_contains_insert [EquivBEq α] [LawfulHashable α] {k a : α} :
+    (m.insert k).contains a → (a == k) = false → m.contains a :=
+  HashMap.contains_of_contains_insert
+
+theorem mem_of_mem_insert [EquivBEq α] [LawfulHashable α] {k a : α} :
+    a ∈ m.insert k → (a == k) = false → a ∈ m :=
+  HashMap.mem_of_mem_insert
+
+@[simp]
+theorem contains_insert_self [EquivBEq α] [LawfulHashable α] {k : α}  : (m.insert k).contains k :=
+  HashMap.contains_insert_self
+
+@[simp]
+theorem mem_insert_self [EquivBEq α] [LawfulHashable α] {k : α} : k ∈ m.insert k :=
+  HashMap.mem_insert_self
+
+@[simp]
+theorem size_empty {c} : (empty c : HashSet α).size = 0 :=
+  HashMap.size_empty
+
+@[simp]
+theorem size_emptyc : (∅ : HashSet α).size = 0 :=
+  HashMap.size_emptyc
+
+theorem isEmpty_eq_size_eq_zero : m.isEmpty = (m.size == 0) :=
+  HashMap.isEmpty_eq_size_eq_zero
+
+theorem size_insert [EquivBEq α] [LawfulHashable α] {k : α} : (m.insert k).size = bif m.contains k then m.size else m.size + 1 :=
+  HashMap.size_insert
+
+theorem size_le_size_insert [EquivBEq α] [LawfulHashable α] {k : α} : m.size ≤ (m.insert k).size :=
+  HashMap.size_le_size_insert
+
+@[simp]
+theorem remove_empty {a : α} {c : Nat} : (empty c : HashSet α).remove a = empty c :=
+  ext HashMap.remove_empty
+
+@[simp]
+theorem remove_emptyc {a : α} : (∅ : HashSet α).remove a = ∅ :=
+  ext HashMap.remove_emptyc
+
+@[simp]
+theorem isEmpty_remove [EquivBEq α] [LawfulHashable α] {k : α} : (m.remove k).isEmpty = (m.isEmpty || (m.size == 1 && m.contains k)) :=
+  HashMap.isEmpty_remove
+
+@[simp]
+theorem contains_remove [EquivBEq α] [LawfulHashable α] {k a : α} : (m.remove k).contains a = (!(a == k) && m.contains a) :=
+  HashMap.contains_remove
+
+@[simp]
+theorem mem_remove [EquivBEq α] [LawfulHashable α] {k a : α} : a ∈ m.remove k ↔ (a == k) = false ∧ a ∈ m :=
+  HashMap.mem_remove
+
+theorem contains_of_contains_remove [EquivBEq α] [LawfulHashable α] {k a : α} : (m.remove k).contains a → m.contains a :=
+  HashMap.contains_of_contains_remove
+
+theorem mem_of_mem_remove [EquivBEq α] [LawfulHashable α] {k a : α} : a ∈ m.remove k → a ∈ m :=
+  HashMap.mem_of_mem_remove
+
+theorem size_remove [EquivBEq α] [LawfulHashable α] {k : α} : (m.remove k).size = bif m.contains k then m.size - 1 else m.size :=
+  HashMap.size_remove
+
+theorem size_remove_le [EquivBEq α] [LawfulHashable α] {k : α} : (m.remove k).size ≤ m.size :=
+  HashMap.size_remove_le
+
+@[simp]
+theorem fst_containsThenInsert {k : α} : (m.containsThenInsert k).1 = m.insert k :=
+  ext HashMap.fst_containsThenInsert
+
+@[simp]
+theorem snd_containsThenInsert {k : α} : (m.containsThenInsert k).2 = m.contains k :=
+  HashMap.snd_containsThenInsert
```

Comment:
I think these should be `containsThenInsert_fst`. It's not entirely sensible, but this seems to be the convention for structure projections.